api/src/main/java/io/grpc/TlsServerCredentials.java

```diff
@@ -0,0 +1,231 @@
+   * enough to use the appropriate APIs to read the configuration. The caller
+   * may support just a subset of a feature, in which case the caller would
+   * need to look at the configuration to determine if only the supported
+   * subset is used.
+   *
+   * <p>This method may not be as simple as a set difference. There may be
+   * multiple features that can independently satisfy a piece of configuration.
+   * If the configuration is incomprehensible, all such features would be
+   * returned, even though only one may be necessary.
+   *
+   * <p>An empty set does not imply that the credentials are fully understood.
+   * There may be optional configuration that can be ignored if not understood.
+   *
+   * <p>Since {@code Feature} is an {@code enum}, {@code understoodFeatures}
+   * should generally be an {@link java.util.EnumSet}. {@code
+   * understoodFeatures} will not be modified.
+   *
+   * @param understoodFeatures the features understood by the caller
+   * @return empty set if the caller can adequately understand the configuration
+   */
+  public Set<Feature> incomprehensible(Set<Feature> understoodFeatures) {
+    Set<Feature> incomprehensible = EnumSet.noneOf(Feature.class);
+    if (fakeFeature) {
+      requiredFeature(understoodFeatures, incomprehensible, Feature.FAKE);
+    }
+    return Collections.unmodifiableSet(incomprehensible);
+  }
+
+  private static void requiredFeature(
+      Set<Feature> understoodFeatures, Set<Feature> incomprehensible, Feature feature) {
+    if (!understoodFeatures.contains(feature)) {
+      incomprehensible.add(feature);
+    }
+  }
+
+  /**
+   * Features to understand TLS configuration. Additional enum values may be added in the future.
+   */
+  public enum Feature {
+    /**
+     * A feature that no consumer should understand. It should be used for unit testing to confirm
+     * a call to {@link #incomprehensible incomprehensible()} is implemented properly.
+     */
+    FAKE,
+    ;
+  }
+
+  /** Creates a builder for changing default configuration. */
+  public static Builder newBuilder() {
+    return new Builder();
+  }
+
+  /** Builder for {@link TlsServerCredentials}. */
+  @ExperimentalApi("TODO")
+  public static final class Builder {
+    private boolean fakeFeature;
+    private byte[] certificateChain;
+    private byte[] privateKey;
+    private String privateKeyPassword;
+
+    private Builder() {}
+
+    /**
+     * Requires {@link Feature#FAKE} to be understood. For use in testing consumers of this
+     * credential.
+     */
+    public Builder requireFakeFeature() {
+      fakeFeature = true;
+      return this;
+    }
+
+    /**
+     * Creates an instance using provided certificate chain and private key. Generally they should
+     * be PEM-encoded and the key is an unencrypted PKCS#8 key (file headers have "BEGIN
+     * CERTIFICATE" and "BEGIN PRIVATE KEY").
+     */
+    public Builder keyManager(File certChain, File privateKey) throws IOException {
```

Comment:
Here you don't actually create/manage `javax.net.ssl.KeyManager` instances so `'keyManager` is a bit misleading but is this to mirror `SslContextBuilder` usage?

Reply:
Yes, it's a mirror. It configures the key manager, at least in an abstract sense. Since there are many ways to configure keys, it seemed the best term to apply to all of them. In the future it seems quite possible there will be a method that allows specifying the `KeyManager` directly.

And yes, this is mirroring `SslContextBuilder`, but it is moreso using the same logic I used when I wrote `SslContextBuilder`.